src/data/polynomial/reverse.lean

```diff
@@ -0,0 +1,202 @@
+  rw [← @rev_at_fun_invol N a, hab, rev_at_fun_invol],
+end
+
+/-- If `i ≤ N`, then `rev_at N i` returns `N - i`, otherwise it returns `i`.
+Essentially, this embedding is only used for `i ≤ N`.
+The advantage of `rev_at N i` over `N - i` is that `rev_at` is an involution.
+-/
+def rev_at (N : ℕ) : function.embedding ℕ ℕ :=
+{ to_fun := λ i , (ite (i ≤ N) (N-i) i),
+  inj' := rev_at_fun_inj }
+
+/-- We prefer to use the bundled `rev_at` over unbundled `rev_at_fun`. -/
+@[simp] lemma rev_at_fun_eq (N i : ℕ) : rev_at_fun N i = rev_at N i := rfl
+
+@[simp] lemma rev_at_invol {N i : ℕ} : (rev_at N) (rev_at N i) = i :=
+rev_at_fun_invol
+
+@[simp] lemma rev_at_le {N i : ℕ} (H : i ≤ N) : rev_at N i = N - i :=
+if_pos H
+
+lemma rev_at_add {N O n o : ℕ} (hn : n ≤ N) (ho : o ≤ O) :
+  rev_at (N + O) (n + o) = rev_at N n + rev_at O o :=
+begin
+  rcases nat.le.dest hn with ⟨n', rfl⟩,
+  rcases nat.le.dest ho with ⟨o', rfl⟩,
+  repeat { rw rev_at_le (le_add_right rfl.le) },
+  rw [add_assoc, add_left_comm n' o, ← add_assoc, rev_at_le (le_add_right rfl.le)],
+  repeat {rw nat.add_sub_cancel_left},
+end
+
+/-- `reflect N f` is the polynomial such that `(reflect N f).coeff i = f.coeff (rev_at N i)`.
+In other words, the terms with exponent `[0, ..., N]` now have exponent `[N, ..., 0]`.
+
+In practice, `reflect` is only used when `N` is at least as large as the degree of `f`.
+
+Eventually, it will be used with `N` exactly equal to the degree of `f`.  -/
+noncomputable def reflect (N : ℕ) (f : polynomial R) : polynomial R :=
+finsupp.emb_domain (rev_at N) f
+
+lemma reflect_support (N : ℕ) (f : polynomial R) :
+  (reflect N f).support = image (rev_at N) f.support :=
+begin
+  ext1,
+  rw [reflect, mem_image, support_emb_domain, mem_map],
+end
+
+@[simp] lemma coeff_reflect (N : ℕ) (f : polynomial R) (i : ℕ) :
+  coeff (reflect N f) i = f.coeff (rev_at N i) :=
+calc finsupp.emb_domain (rev_at N) f i
+    = finsupp.emb_domain (rev_at N) f (rev_at N (rev_at N i)) : by rw rev_at_invol
+... = f.coeff (rev_at N i) : finsupp.emb_domain_apply _ _ _
+
+@[simp] lemma reflect_zero {N : ℕ} : reflect N (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_eq_zero_iff {N : ℕ} {f : polynomial R} :
+  reflect N (f : polynomial R) = 0 ↔ f = 0 :=
+begin
+  split,
+  { intros a,
+    injection a with f0 f1,
+    rwa [map_eq_empty, support_eq_empty] at f0, },
+  { rintro rfl,
+    refl, },
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (N : ℕ) :
+  reflect N (f + g) = reflect N f + reflect N g :=
+by { ext1, rw [coeff_add, coeff_reflect, coeff_reflect, coeff_reflect, coeff_add], }
+
+@[simp] lemma reflect_C_mul (f : polynomial R) (r : R) (N : ℕ) :
+  reflect N (C r * f) = C r * (reflect N f) :=
+by { ext1, rw [coeff_reflect, coeff_C_mul, coeff_C_mul, coeff_reflect], }
+
+@[simp] lemma reflect_C_mul_X_pow (N n : ℕ) {c : R} :
+  reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext,
+  rw [reflect_C_mul, coeff_C_mul, coeff_C_mul, coeff_X_pow, coeff_reflect],
```

Comment:
```suggestion
  simp only [reflect_C_mul, coeff_C_mul, coeff_X_pow, coeff_reflect],
```

Reply:
@Vierkantor , @jcommelin 
I reverted this ```simp only``` to a ```rw```: this is what caused the CI failure.  It was the only "non-final" ```simp``` and the rest of the proof depended on the exact output of the ```rw```: ```simp only``` went too far!

Reply:
Ah, oops! I must have overlooked the error when I tried it out.

Reply:
No problem!  As always, thank you for your help!